### ISSUE TITLE
MixedReality namespace not found when using OpenXR for Android

### ISSUE THREAD

**Issue by CZSeas** [Azure/azure-spatial-anchors-samples#283]
**Description**

Cannot compile when building for Android due to SpatialAnchorExtensions not locating the Microsoft.MixedReality namespace. MRTK 2.7.2 installed and working. Building for UWP works as intended. Imported ASA using MR feature tool.

**Steps to reproduce the issue**

1. Switch build platform to Android
2. Build

**Screenshots**

![image](https://user-images.githubusercontent.com/54338147/129266080-8c163a44-3d06-4659-9e51-2e5af7247d2c.png)

**Development information (please complete the following information)**

 - OS: Windows
 - Development Platform\Version: Unity 2020.3.11f1
 - Azure Spatial Anchors SDK version: 2.10.2

**Comment by wyliefoxxx** [comment]
Crazy timing.. I've been doing my darndest to upgrade our project to use the new ASA version 2.10.2 all day, and have been dealing with the same issue upon build attempts:

"Library\PackageCache\com.microsoft.azure.spatial-anchors-sdk.core@715ff388952b-1628106712006\Runtime\Scripts\SpatialAnchorExtensions.cs(17,17): error CS0234: The type or namespace name 'MixedReality' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?)"

The referenced lines in SpatialAnchorExtensions:
#if MIXED_REALITY_OPENXR // version define for optional Microsoft Mixed Reality OpenXR package
using Microsoft.MixedReality.OpenXR;
#endif

Previously we were using:
Unity 2019.4.8.f1
MRTK 2.4
ASA 2.4

After going through (quite a few) number of upgrade steps, including deleting all MRTK files and upgrading with the new MixedRealityFeatureTool (love it btw), we are using:
Unity 2020.3.15f2
MRTK 2.7.2
ASA 2.10.2
Mixed Reality OpenXR Plugin 1.0.1
![image](https://user-images.githubusercontent.com/28640699/129283645-61466754-e51e-46e7-b873-ac81b9f3ec3d.png)

I've tried removing all related packages, and reimporting them in a number of ways, along with deleting my 'Library' folder, (in different orders), 'Reset Packages to Default', and everything else I could think of to resolve this issue. The project runs fine in the editor, but fails quickly upon build with the error above.

We're super excited to push our project forward, and utilize what the release notes say is 'a change that improves the accuracy of spatial anchors', but we're a little dead in the water until we can get past this build error. I'll probably be trying different combinations of the previous versions of the different packages until I can get one to work with ASA 2.10.0

Any advice, direction, or options would be great at this point. Thanks.

**Comment by wyliefoxxx** [comment]
Seems like a similar issue to this one:
https://github.com/Azure/azure-spatial-anchors-samples/issues/124

**Comment by seancheno** [comment]
We're also trying to upgrade to ASA 2.10.2 on Android with Unity and getting the same error. Any help is hugely appreciated.

**Comment by vastja** [comment]
We have encountered the same issue in Unity 2020.3.15f2 for ASA 2.10.2, MRTK 2.7.2, microsoft.mixedreality.openxr-1.0.2.

It turned out the problem was indeed in Microsoft.MixedReality.OpenXR, because its **asmdef** file was not configured for Android. Unfortunately you can't just change this file if it is located in Library/PackageCache (this should be the location if you use package manager or manifest.json). 
But there is a workaround: 

1. Move this library (com.microsoft.mixedreality.openxr) to **Packages** folder (this will enable persistent asmdef modifications)
2. Set this library (modify **asmdef**) to be included for Android (see images below)

![explorer](https://user-images.githubusercontent.com/22541583/129345364-7844a848-a196-4c4e-8826-12657a068f04.PNG)

![asmdef](https://user-images.githubusercontent.com/22541583/129345341-6e5b49ee-8f48-4ee1-82fe-758860711b87.PNG)

**Comment by wyliefoxxx** [comment]
Thank you very much for that workaround Vastja! I can confirm that this workaround fixed the issue, and our project is now building for android. Thanks for the timely input!

**Comment by oafolabi-msft** [comment]
Thanks for your input @wyliefoxxx, @seancheno and @vastja. @CZSeas , did the workaround resolve the issue for you?

**Comment by vastja** [comment]
Hi I found out the problem is more complex than I thought it was at first. If you just use my first workaround it may happen that **SpatialAnchorManager.StartSessionAsync()** won't work correctly (see [https://github.com/Azure/azure-spatial-anchors-samples/issues/241](#241).

The problem is in **MIXED_REALITY_OPENXR preprocessor directive**. There is either some bug in Android ASA implementation, or it is badly designed. As you can see in _com.microsoft.azure.spatial-anchors-sdk.core_ this directive is defined if com.microsoft.mixedreality.openxr package is present in the project (see attached image) (in our case it is present because we would like to use it for UWP). Unfortunately if this directive is present in Android build **SpatialAnchorManager.StartSessionAsync()** does not work correctly. @oafolabi-msft can you investigate what is wrong? Can it be that on Android you will simply use Unity's AR Foundation and this directive force ASA to use OpenXR (which is probably not set for Android)?

**I managed to find another workaround:**

1. Move this library (com.microsoft.azure.spatial-anchors-sdk.core) to Packages folder (this will enable persistent asmdef modifications)
2. Remove MIXED_REALITY_OPENXR directive form **asmdef** (see image below)
3. If you want to use OpenXR for UWP (as we do). It should be enough to set preprocessor directive in Player > tab Other Settings > Scripting Define Symbols (make sure to set it for correct target platform)

![mixed-reality-openxr](https://user-images.githubusercontent.com/22541583/129553259-7ef4e598-606b-4e1d-9786-9a6c316045dd.PNG)

**Comment by wyliefoxxx** [comment]
Thanks again @vastja! I can confirm that this last workaround has brought our system back up and running fully. (see https://github.com/Azure/azure-spatial-anchors-samples/issues/241) The walkthrough helps immensely. It seems as though ASA 2.10.2 is now working well on Android, and, after preliminary testing, I do see a marked improvement in spatial anchor accuracy, which is very exciting (Thank you to the development team!) Deeper into the metaverse we go! ;)

(I would love to hear a little bit more about this issue once it's resolved fully - I'm sure we'll see a fix in upcoming versions, but with these workarounds, are we really taking advantage of ASA's new capacity to the fullest on Android?)

**Comment by oafolabi-msft** [comment]
Hi @vastja,  thanks for letting us know! We're looking into this (internal issue #35362481) and will update you with progress. Thanks!"

**Comment by msftradford** [comment]
Thanks @wyliefoxxx and @vastja for the investigation and explanation. Very happy to hear that the workaround has unblocked you and that you can see accuracy improvements with our latest release! We are working on a fix to this issue that will be included as a part of our next SDK release. @wyliefoxxx these workarounds should not affect our latest improvements to the ASA Android SDK.

**Comment by CZSeas** [comment]
Thanks @wyliefoxxx, @vastja for the workaround, seems to be working for my project!

**Comment by MehranAzimi-msft** [comment]
As noted above, the workarounds posted above by @wyliefoxxx and @vastja both works. Thanks for posting those. 

I like to add that you can workaround this using the following method as well. Similar to other workaround, this method requires editing the content of com.microsoft.azure.spatial-anchors-sdk.core package. As noted, if this package is included in you project via scoped registry, then editing the plugin code will not work as Unity will overwrite your changes as you build. You need to download the com.microsoft.azure.spatial-anchors-sdk.core from [here](https://dev.azure.com/aipmr/MixedReality-Unity-Packages/_packaging?_a=package&feed=Unity-packages&package=com.microsoft.azure.spatial-anchors-sdk.core&protocolType=Npm&version=2.10.2), expand the tgz content into a folder, and then import the package to your project using [these instruction](https://docs.unity3d.com/Manual/upm-ui-local.html). Then edit the 
Runtime\Scripts\SpatialAnchorManager.cs  and SpatialAnchorExtensions.cs files inside ASA package, and replace instances of 
`#if MIXED_REALITY_OPENXR`
with 
`#if (UNITY_WSA || WINDOWS_UWP) && MIXED_REALITY_OPENXR`

As @msftradford stated, ASA team will include a fix for this in next ASA SDK release.

**Comment by sana2024** [comment]
I have the same issues, it works %of the time on one device or two but when I try to start the session on multiple devices it gives me the same error, I have to click the start button multiple times to make it work, and this issues also occurs when I try to change scenes. please help.

**Comment by darax** [comment]
Hi @sna2024, the above error is in reference to building the project, your short description sounds like a problem after the project is built. Please open a new issue with details of what you are facing, including logs if you have them.
